src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -104,10 +104,20 @@ std::shared_ptr<RayObject> GetRequest::Get(const ObjectID &object_id) const {
   return nullptr;
 }
 
-CoreWorkerMemoryStore::CoreWorkerMemoryStore() {}
+CoreWorkerMemoryStore::CoreWorkerMemoryStore(std::shared_ptr<ReferenceCounter> counter) {
+  if (counter != nullptr) {
+    ref_counter_ = counter;
+    ref_counter_->OnObjectDeleted([this](const ObjectID &obj_id) { Delete({obj_id}); });
+  }
+}
 
 Status CoreWorkerMemoryStore::Put(const ObjectID &object_id, const RayObject &object) {
   std::unique_lock<std::mutex> lock(lock_);
+
+  if (ref_counter_ && !ref_counter_->HasReference(object_id)) {
```

Comment:
Ahh is this an optimization for if the return value is out of scope? Nice. Can you add a comment?

Reply:
Done

Reply:
@ericl I might miss something here - if an object ID is already out-of-scope, how would user code invoke a Put in this case?